### ISSUE TITLE
matrix-js-sdk no longer works in node

### ISSUE THREAD

**Issue by richvdh** [matrix-org/matrix-js-sdk#4287]
https://github.com/matrix-org/matrix-js-sdk/issues/4284 describes a problem wherein matrix-js-sdk no longer works in a CommonJS environment, but actually it's worse than that: it doesn't work in an ES modules environment either.

If we apply the fix from https://github.com/matrix-org/matrix-js-sdk/pull/4285, and run in an esm environment, I get:

```
node:internal/modules/esm/resolve:260
    throw new ERR_MODULE_NOT_FOUND(
          ^

Error [ERR_MODULE_NOT_FOUND]: Cannot find module '/home/rav/work/matrix-js-sdk/examples/node/node_modules/matrix-js-sdk/lib/matrix' imported from /home/rav/work/matrix-js-sdk/examples/node/node_modules/matrix-js-sdk/lib/index.js
    at finalizeResolution (node:internal/modules/esm/resolve:260:11)
    at moduleResolve (node:internal/modules/esm/resolve:920:10)
    at defaultResolve (node:internal/modules/esm/resolve:1119:11)
    at ModuleLoader.defaultResolve (node:internal/modules/esm/loader:542:12)
    at ModuleLoader.resolve (node:internal/modules/esm/loader:511:25)
    at ModuleLoader.getModuleJob (node:internal/modules/esm/loader:241:38)
    at ModuleJob._link (node:internal/modules/esm/module_job:126:49) {
  code: 'ERR_MODULE_NOT_FOUND',
  url: 'file:///home/rav/work/matrix-js-sdk/examples/node/node_modules/matrix-js-sdk/lib/matrix'
}

Node.js v22.2.0
```

The problem appears to be that `index.js` contains:
```js
import * as matrixcs from "./matrix";
```

However, the file it actually wants is `./matrix.js`, and the nodejs ESM resolver doesn't know it needs to add the extension. Likewise for every link in the project :/

**Comment by richvdh** [comment]
Apparently, the typescript way to do this is actually to

 ```ts
import {foo} from "./matrix.js"`; // note js extension
```

The `.js` is then retained in the output, but typescript is intelligent enough to use the ts file for type definitions (cf https://www.typescriptlang.org/docs/handbook/modules/reference.html#file-extension-substitution)

**Comment by 0ctoPusK** [comment]
I got the same error as you had

**Comment by richvdh** [comment]
If anybody would like to go through all the `import` statements in the js-sdk source and add the `.js` suffix to all the filenames, I'd happily review a pull-request.

**Comment by Johennes** [comment]
Bit of a wild idea but maybe we could just tweak the imports with babel? I haven't tested this a lot but https://github.com/matrix-org/matrix-js-sdk/pull/4373 at least lets me run the Node.js example locally.

**Comment by Johennes** [comment]
Alternatively, #4375 script-replaces all imports. This also lets me run the Node.js example but appears to break most test.

**Comment by richvdh** [comment]
SIGH. How can this possibly be so difficult? I can't believe we are the first library ever to use typescript and jest, and transpile to ES modules.

Maybe what we want is to have the source say `import {foo} from "./matrix.ts"` and enable https://babeljs.io/docs/babel-preset-typescript#rewriteimportextensions, except when `process.env.NODE_ENV === "test"` ?

**Comment by Johennes** [comment]
> Maybe what we want is to have the source say `import {foo} from "./matrix.ts"` and enable https://babeljs.io/docs/babel-preset-typescript#rewriteimportextensions, except when `process.env.NODE_ENV === "test"` ?

This looks promising: https://github.com/matrix-org/matrix-js-sdk/pull/4377

**Comment by ledlamp** [comment]
omg. How do I use this library then???

**Comment by Johennes** [comment]
You might be able to tweak imports by using a bundler in your project until #4377 lands and is released.

**Comment by richvdh** [comment]
This should be fixed in v34.5.
